src/main/java/net/neoforged/neoforge/client/util/TooltipUtil.java

```diff
@@ -0,0 +1,349 @@
+     * <p>
+     * After the tooltip lines have been added, fires the {@link AddAttributeTooltipsEvent} to allow mods to add additional attribute-related lines.
+     * 
+     * @param tooltip A consumer to add the tooltip lines to.
+     * @param ctx     The tooltip context
+     */
+    public static void addAttributeTooltips(ItemStack stack, Consumer<Component> tooltip, AttributeTooltipContext ctx) {
+        ItemAttributeModifiers modifiers = stack.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
+        if (modifiers.showInTooltip()) {
+            applyModifierTooltips(stack, tooltip, ctx);
+        }
+        NeoForge.EVENT_BUS.post(new AddAttributeTooltipsEvent(stack, tooltip, ctx));
+    }
+
+    /**
+     * Unconditionally applies the attribute modifier tooltips for all attribute modifiers present on the item stack.
+     * <p>
+     * Before application, this method posts the {@link GatherSkippedAttributeTooltipsEvent} to determine which tooltips should be skipped.
+     * 
+     * @param stack
+     * @param tooltip
+     * @param ctx
+     */
+    public static void applyModifierTooltips(ItemStack stack, Consumer<Component> tooltip, AttributeTooltipContext ctx) {
+        Set<ResourceLocation> skips = new HashSet<>();
+        var event = NeoForge.EVENT_BUS.post(new GatherSkippedAttributeTooltipsEvent(stack, skips, ctx));
+        if (event.isSkippingAll()) {
+            return;
+        }
+
+        for (EquipmentSlotGroup group : EquipmentSlotGroup.values()) {
+            Multimap<Holder<Attribute>, AttributeModifier> modifiers = AttributeUtil.getSortedModifiers(stack, group);
+            Component groupName = Component.translatable("item.modifiers." + group.getSerializedName()).withStyle(ChatFormatting.GRAY);
+            applyTextFor(stack, tooltip, modifiers, groupName, skips, ctx);
+        }
+    }
+
+    /**
+     * Applies the text for a single group of modifiers to the tooltip for a given item stack.
+     * <p>
+     * This method will attempt to merge multiple modifiers for a single attribute into a single modifier if {@linkplain NeoForgeMod#enableMergedAttributeTooltips()} was called.
+     * 
+     * @param stack       The item stack that owns the modifiers
+     * @param tooltip     The consumer to append tooltip components to
+     * @param modifierMap A mutable map of modifiers for the given group
+     * @param groupName   The name of the current modifier group
+     * @param skips       A set of modifier IDs to not apply to the tooltip
+     * @param ctx         The tooltip context
+     */
+    public static void applyTextFor(ItemStack stack, Consumer<Component> tooltip, Multimap<Holder<Attribute>, AttributeModifier> modifierMap, Component groupName, Set<ResourceLocation> skips, AttributeTooltipContext ctx) {
```

Comment:
Since we've noted that the tooltip code has to move to common anyway, I may just flatten this class with `AttributeUtil` (which then makes the purpose clearer).  The method names might still need a slight rework though, as `AttributeUtil.applyTextFor` isn't exactly the best of names.